### ISSUE TITLE
Language not changing

### ISSUE THREAD

**Issue by mohamed-elmarakby** [aissat/easy_localization#448]
I've went through other issues and this isn't like #370.

Application changes the direction of the language and the locale gets changed but the words of the application don't get changed they remain on the same language until I restart the application myself, it was working fine on the ^2.3.3 version before null safety and it changed instantly now I don't know what happened!

I've tried it on a normal application, using provider, and using bloc and it's the same issue

**Comment by 11dj** [comment]
I got same issue too.
In my case I want to change language from model class by passing context from StatefulWidget to Model class to call context.locale. And it doesn't work.

**Comment by aissat** [comment]
Please more details ??!

**Comment by kechankrisna** [comment]
Please test with MaterialApp.router, you will see the language not refresh in UI when we change the language. It works fine in MaterialApp but not MaterialApp.router

**Comment by Faaatman** [comment]
The solution that worked for me is the following:
My app has a bottom navigation bar where the tabs are in `IndexedStack` I was supplying the `IndexedStack` children list with `const`
```dart
children: [
  const ProfilePage();
  const HomePage();
]
```
The fix is to remove the `const` as follows:
```dart
children: [
    // ignore: prefer_const_constructors
   ProfilePage();
    // ignore: prefer_const_constructors
   HomePage();
]
```
Every other widget inside HomePage or ProfilePage can be `const`.

**Comment by burakkurtarir** [comment]
> The solution that worked for me is the following: My app has a bottom navigation bar where the tabs are in `IndexedStack` I was supplying the `IndexedStack` children list with `const`
> 
> ```dart
> children: [
>   const ProfilePage();
>   const HomePage();
> ]
> ```
> 
> The fix is to remove the `const` as follows:
> 
> ```dart
> children: [
>     // ignore: prefer_const_constructors
>    ProfilePage();
>     // ignore: prefer_const_constructors
>    HomePage();
> ]
> ```
> 
> Every other widget inside HomePage or ProfilePage can be `const`.

Yes you are awesome, saved my day, thank you.

**Comment by tijanirf** [comment]
> The solution that worked for me is the following: My app has a bottom navigation bar where the tabs are in `IndexedStack` I was supplying the `IndexedStack` children list with `const`
> 
> ```dart
> children: [
>   const ProfilePage();
>   const HomePage();
> ]
> ```
> 
> The fix is to remove the `const` as follows:
> 
> ```dart
> children: [
>     // ignore: prefer_const_constructors
>    ProfilePage();
>     // ignore: prefer_const_constructors
>    HomePage();
> ]
> ```
> 
> Every other widget inside HomePage or ProfilePage can be `const`.

Do we have better workaround so we don't need to remove the `const` keyword? Yes, i am aware of what `const` does but it is kind of hard to slice the widget if it is so deep in the widget tree. With current default linter settings, it's kind of odd to ignore it manually as it is not intuitive at least for me.

**Comment by AlexanderFarkas** [comment]
Easy Localization uses bad practice - it doesn't utilize BuildContext. That's why your translations will never be updated if they are const.

**Comment by Harits19** [comment]
> > The solution that worked for me is the following: My app has a bottom navigation bar where the tabs are in `IndexedStack` I was supplying the `IndexedStack` children list with `const`
> > ```dart
> > children: [
> >   const ProfilePage();
> >   const HomePage();
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The fix is to remove the `const` as follows:
> > ```dart
> > children: [
> >     // ignore: prefer_const_constructors
> >    ProfilePage();
> >     // ignore: prefer_const_constructors
> >    HomePage();
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Every other widget inside HomePage or ProfilePage can be `const`.
> 
> Do we have better workaround so we don't need to remove the `const` keyword? Yes, i am aware of what `const` does but it is kind of hard to slice the widget if it is so deep in the widget tree. With current default linter settings, it's kind of odd to ignore it manually as it is not intuitive at least for me.

don't forget to delete the 'const' on MaterialApp child/home widget if it exists

**Comment by mauryagaurav947** [comment]
Same issue context.setLocale does not effect. Please help.

**Comment by andreamainella98** [comment]
To update the translations, since Easy_Localization doesn't use the context, just update the context of the EasyLocalization instance.

With this solution I managed to fix this problem.

My Code:
```dart
Padding(
  padding: const EdgeInsets.all(18.0),
  child: DropdownButton(
    items: context.supportedLocales
        .map(
          (e) => DropdownMenuItem(
            value: e,
            child: Text(
              context.localizables.localeTitle(e.languageCode),
              style: context.styles.h2Lato,
            ),
          ),
        )
        .toList(),
    value: context.locale,
    onChanged: (value) => value != null ? context.setLocale(value) : null,
  ),
),
```
```dart
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/material.dart';

extension LanguageLoader on BuildContext {
  Localizable get localizables => Localizable(this);
}

class Localizable {
  final BuildContext context;

  Localizable(this.context);

  String get appTitle => ltr('appTitle');

  String get yes => ltr('yes');

  String get no => ltr('no');

  String get quoteAnime => ltr('quoteAnime');

  String darkModeTitle(bool enabled) => ltr('darkModeTitle', args: [enabled ? yes : no]);

  String localeTitle(String locale) => ltr('locale.$locale');

  String ltr(
    String key, {
    List<String>? args,
    Map<String, String>? namedArgs,
    String? gender,
  }) {
    EasyLocalization.of(context);
    return tr(key, args: args, namedArgs: namedArgs, gender: gender);
  }
}

```

**Comment by aissat** [comment]
Duplicate of #552

**Comment by ilyasarafathemvigo** [comment]
I have the same issue

**Comment by LiamMarega** [comment]
same issue here

**Comment by andreamainella98** [comment]
@ilyasarafathemvigo @LiamMarega look my solution

> To update the translations, since Easy_Localization doesn't use the context, just update the context of the EasyLocalization instance.
> 
> With this solution I managed to fix this problem.
> 
> My Code:
> 
> ```dart
> Padding(
>   padding: const EdgeInsets.all(18.0),
>   child: DropdownButton(
>     items: context.supportedLocales
>         .map(
>           (e) => DropdownMenuItem(
>             value: e,
>             child: Text(
>               context.localizables.localeTitle(e.languageCode),
>               style: context.styles.h2Lato,
>             ),
>           ),
>         )
>         .toList(),
>     value: context.locale,
>     onChanged: (value) => value != null ? context.setLocale(value) : null,
>   ),
> ),
> ```
> 
> ```dart
> import 'package:easy_localization/easy_localization.dart';
> import 'package:flutter/material.dart';
> 
> extension LanguageLoader on BuildContext {
>   Localizable get localizables => Localizable(this);
> }
> 
> class Localizable {
>   final BuildContext context;
> 
>   Localizable(this.context);
> 
>   String get appTitle => ltr('appTitle');
> 
>   String get yes => ltr('yes');
> 
>   String get no => ltr('no');
> 
>   String get quoteAnime => ltr('quoteAnime');
> 
>   String darkModeTitle(bool enabled) => ltr('darkModeTitle', args: [enabled ? yes : no]);
> 
>   String localeTitle(String locale) => ltr('locale.$locale');
> 
>   String ltr(
>     String key, {
>     List<String>? args,
>     Map<String, String>? namedArgs,
>     String? gender,
>   }) {
>     EasyLocalization.of(context);
>     return tr(key, args: args, namedArgs: namedArgs, gender: gender);
>   }
> }
> ```

**Comment by Margub-Morshed** [comment]
In my flutter project, instead of "GetMaterialApp" use "MaterialApp" saved me. Now dynamically instant changed locale showed. Hope someone find it useful.

**Comment by antiv** [comment]
Maybe stupid solutution but I add `key: Key('Bottom Navigation ${context.locale}'),` to my bottom navigation widget to fix this problem :)

**Comment by Gerry1218** [comment]
> In my flutter project, instead of "GetMaterialApp" use "MaterialApp" saved me. Now dynamically instant changed locale showed. Hope someone find it useful.

I want to use both？ How to resolve？

**Comment by sofasurfa** [comment]
@Faaatman 's solution worked for me as well. It easy to overlook this by mistake. In my case I was changing my Locale to Arabic - which caused Text widgets to change direction (right to left), but text was still in English... I spent hours debugging my [redux state management](https://github.com/marcglasberg/async_redux) thinking there was something wrong.. 

But all I had to do was to remove **const** in my `Navigator` route....  Guess it just one of these "Flutter gotchas" 😅
@aissat maybe we could document this somewhere as "Gotchas" at the end of pkg documentation?

```
case OuterRoutes.search:
                  return MaterialPageRoute(
                    builder: (context) => const SearchPage(), // --> remove "const" so widget rebuilds
    );
```
